HardGLU implementation

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42443 [quant] quantized GLU and HardGLU
* **#42728 HardGLU implementation**
* #42727 Factoring out common code in the GLU
* #42442 [quant] Quantization parameters in TensorIterator

Differential Revision: [D23011805](https://our.internmc.facebook.com/intern/diff/D23011805)

Commit:
HardGLU implementation

[ghstack-poisoned]

z-a-f commented on torch/nn/functional.py:
```diff
@@ -1716,11 +1745,11 @@ def silu(input, inplace=False):
-        `Sigmoid-Weighted Linear Units for Neural Network Function Approximation 
-        in Reinforcement Learning <https://arxiv.org/abs/1702.03118>`_ and `Swish: 
-        a Self-Gated Activation Function <https://arxiv.org/abs/1710.05941v1>`_ 
+        See `Gaussian Error Linear Units (GELUs) <https://arxiv.org/abs/1606.08415>`_
```
Whoops -- auto linter thingy....

vkuzo commented on test/test_torch.py:
```diff
@@ -16581,6 +16581,30 @@ def test_hardsigmoid(self, device, dtype):
 
+    @dtypes(torch.float, torch.double)
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_hardglu(self, device, dtype):
```
can we also check the derivative?  `torch.autograd.gradcheck` could be useful

vkuzo reviewed (commented):
can we add an op benchmark as well?

Commit:
Update on "HardGLU implementation"




[ghstack-poisoned]

Commit:
Update on "HardGLU implementation"


Differential Revision: [D23011805](https://our.internmc.facebook.com/intern/diff/D23011805)

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/native/cuda/Activation.cu:
```diff
@@ -478,6 +478,36 @@ void hardsigmoid_backward_kernel(TensorIterator& iter) {
 }
 
+void hardglu_kernel(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, iter.dtype(), "hardsigmoid_cuda", [&]() {
```
NIT: Typo here "hardsigmoid_cuda"?

raghuramank100 commented on aten/src/ATen/native/GatedLinearUnit.cpp:
```diff
@@ -86,5 +88,51 @@ Tensor glu_backward(const Tensor& grad_output, const Tensor& input, int64_t dim)
+  return at::hardglu_out(result, self, dim);
+}
+
+Tensor& hardglu_backward_out(Tensor& grad_input,
```
Not sure if I follow the math here. The gradient for hard GLU has two components:
HGLU(x) = x[0:n-1].Hsigmoid([x[n:end]) 
gradient[i] = HS[i+n]  * y[i] if i <n
 else: 
gradient[i+N] = x[0:n-1].Gradient(Hsigmoid). 
Is this what is being done by the code below?

Commit:
Update on "HardGLU implementation"


Differential Revision: [D23011805](https://our.internmc.facebook.com/intern/diff/D23011805)

[ghstack-poisoned]

Commit:
Update on "HardGLU implementation"


Differential Revision: [D23011805](https://our.internmc.facebook.com/intern/diff/D23011805)

[ghstack-poisoned]
